### ISSUE TITLE
connection lock forever when losting connection

### ISSUE THREAD

**Issue by seven1240** [Eonblast/Emysql#7]
I tested a bit of connection-auto-recover when lost connection with only one pool

First I run a query and make sure it works, then stop mysql, and start, then run execute results one exception, but following executes get recovered as expected. 

1) emysql:execute(mysql2, "select 1").
*\* exception exit: {failed_to_recv_packet_header,closed}
     in function  emysql:monitor_work/3

2) emysql:execute(mysql2, "select 1").
{result_packet,5,
               [{field,2,<<"def">>,<<>>,<<>>,<<>>,<<"1">>,<<>>,8,<<>>,63,1,
                       129,0}],
               [[1]],
               <<>>}

Then, I closed mysql again,  and execute returns econnrefused

1) emysql:execute(mysql2, "select 1").
*\* exception exit: {failed_to_connect_to_database,econnrefused}
     in function  emysql_conn:open_connection/1
     in call from emysql_conn:reset_connection/2
     in call from emysql:monitor_work/3

Then I started mysql again, all following executes returns connection_lock_timeout and the connection cannot be recovered.

2) emysql:execute(mysql2, "select 1").
*\* exception exit: connection_lock_timeout
     in function  emysql_conn_mgr:wait_for_connection/1
     in call from emysql:execute/4

So I guess the lock should be released on the 1) case of econnrefused etc.  

I haven't test other error cases.

Thanks.

**Comment by Eonblast** [comment]
I see what you mean but is this a practical concern or rather theoretical?

I am not sure how the actual order of things would be in production. Server going down, coming back up and queries being repeated. If the server is still down during the first repeat, the connection never recovers. All connections would probably go bad. So your sample is of a connection that is idle during which window the server goes down and comes back up. But is that a real scenario or is the actual challenge the first retry while the server is sill unresponsive.

Still trying to get my head around it, let me know what your thoughts are,

thanks,
Henning

**Comment by seven1240** [comment]
Let's say we use erlang for a WEB server, and each request is a new process and when DB down, all processes would just fail. but after DB back in service, the WEB server should auto recover the connections to DB and get back as normal. Makes sense?

**Comment by brendonh** [comment]
This just happened to us overnight, in the way seven1240 describes above, effectively killing the whole server. What's the appropriate workaround?

**Comment by Eonblast** [comment]
Are you sure it's issue #7, or could it be #9?

Thanks,
Henning

**Comment by Eonblast** [comment]
I found it, the renew and reset are flawed in that they try to connect while not in a separate monitored process. This can have the process dying w/o a chance to unlock the connection before it goes.

**Comment by Eonblast** [comment]
Proposed fix in branch 'issue7'. 

The changes are not minimal but straight forward.

https://github.com/Eonblast/Emysql/tree/issue7

Full Diff: https://github.com/Eonblast/Emysql/compare/master...issue7

Basically, I introduced a catch around the renew/reset of connections to make sure that they end up unlocked. I also put 'dead' connections back into the queue to make sure it does not run dry during the time that the sql server is down. I unified "renewed" and reset for this and differentiated by parameter.

There is another branch that does the same and traces extensively, which can help understanding what goes on: https://github.com/Eonblast/Emysql/tree/issue7traces
